flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/BatchShuffleReadBufferPool.java

```diff
@@ -0,0 +1,353 @@
+                                        + " should be increased.",
+                                allocated * bufferSize,
+                                (numTotalBuffers - allocated) * bufferSize,
+                                TaskManagerOptions.FRAMEWORK_OFF_HEAP_MEMORY.key(),
+                                TaskManagerOptions.NETWORK_BATCH_SHUFFLE_READ_MEMORY.key(),
+                                TaskManagerOptions.FRAMEWORK_OFF_HEAP_MEMORY.key(),
+                                TaskManagerOptions.TASK_OFF_HEAP_MEMORY.key()));
+            }
+        }
+    }
+
+    /**
+     * Requests a collection of buffers (determined by {@link #numBuffersPerRequest}) from this
+     * buffer pool. Exception will be thrown if no enough buffers can be allocated in the given
+     * timeout and the corresponding requester is not holding any allocated buffers currently.
+     */
+    public List<MemorySegment> requestBuffers(Object owner) throws Exception {
+        checkArgument(owner != null, "Owner must be not null.");
+
+        List<MemorySegment> allocated = new ArrayList<>(numBuffersPerRequest);
+        synchronized (buffers) {
+            checkState(!destroyed, "Buffer pool is already destroyed.");
+
+            if (!initialized) {
+                initialize();
+            }
+
+            long startTime = System.currentTimeMillis();
+            Counter counter = numBuffersAllocated.get(owner);
+            while (buffers.size() < numBuffersPerRequest
+                    || (counter != null
+                            && counter.get() + numBuffersPerRequest > maxBuffersPerRequester)) {
+                checkState(!destroyed, "Buffer pool is already destroyed.");
+
+                buffers.wait(requestTimeout);
+                counter = numBuffersAllocated.get(owner);
+
+                if (counter == null && System.currentTimeMillis() - startTime >= requestTimeout) {
+                    throw new TimeoutException(
+                            String.format(
+                                    "Can't allocate enough buffers in the given timeout, which means"
+                                            + " there is a fierce contention for read buffers, please"
+                                            + " increase %s.",
+                                    TaskManagerOptions.NETWORK_BATCH_SHUFFLE_READ_MEMORY.key()));
+                }
+            }
+
+            while (allocated.size() < numBuffersPerRequest) {
+                allocated.add(buffers.poll());
+            }
+
+            if (counter == null) {
+                counter = new Counter();
+                numBuffersAllocated.put(owner, counter);
+            }
+            counter.increase(allocated.size());
+        }
+        return allocated;
+    }
+
+    /**
+     * Recycles the target buffer to this buffer pool. This method should never throw any exception.
+     */
+    public void recycle(MemorySegment segment, Object owner) {
```

Comment:
Could we proxy this method to `recycle(Collections.singletonList(segment), owner)` ?